### PR TITLE
gh-115685: Type/values propagate for TO_BOOL in tier 2

### PR DESCRIPTION
**Pull request by Fidget-Spinner**
* Issue: gh-115685

### CODE REVIEW

File: Python/tier2_redundancy_eliminator_bytecodes.c
Comment (markshannon) [inline Python/tier2_redundancy_eliminator_bytecodes.c]: This is wrong. `value` is most likely not a boolean.

File: Python/tier2_redundancy_eliminator_bytecodes.c
Comment (markshannon) [inline Python/tier2_redundancy_eliminator_bytecodes.c]: This seems incorrect, there is no reason why the constant needs to be immortal.

File: Python/tier2_redundancy_eliminator_bytecodes.c
Comment (markshannon) [inline Python/tier2_redundancy_eliminator_bytecodes.c]: This is also wrong. lists are mutable.

**Review by markshannon — CHANGES_REQUESTED** [review]
All the `res = sym_new_...` assignments need to wrapped in `OUT_OF_SPACE_IF_NULL`

File: Python/tier2_redundancy_eliminator_bytecodes.c
Comment (Fidget-Spinner) [reply Python/tier2_redundancy_eliminator_bytecodes.c]: Woops I misunderstood what `PyObject_IsTrue` does. Thanks.!

**Comment by Fidget-Spinner** [comment]
I have made the requested changes; please review again.

**Comment by gvanrossum** [comment]
Odd. That test is still failing?!

File: Python/tier2_redundancy_eliminator_bytecodes.c
Comment (markshannon) [inline Python/tier2_redundancy_eliminator_bytecodes.c]: This could call arbitrary code, so it is not safe to call here, and `value` might be mutable so it is incorrect.

File: Python/tier2_redundancy_eliminator_bytecodes.c
Comment (markshannon) [inline Python/tier2_redundancy_eliminator_bytecodes.c]: If `value` is already a `bool` then this can become a `_NOP`

File: Python/tier2_redundancy_eliminator_bytecodes.c
Comment (markshannon) [inline Python/tier2_redundancy_eliminator_bytecodes.c]: This should be converted to `_POP_TOP; LOAD_CONST_BORROW`. You will need to
* add a `_POP_TOP_LOAD_CONST_BORROW` super-micro-op, or
* in a preparatory issue, add machinery to allow efficient inserting of multiple micro-ops during optimization.

File: Python/tier2_redundancy_eliminator_bytecodes.c
Comment (markshannon) [inline Python/tier2_redundancy_eliminator_bytecodes.c]: See earlier comment about `_POP_TOP; LOAD_CONST_BORROW`

File: Python/tier2_redundancy_eliminator_bytecodes.c
Comment (markshannon) [inline Python/tier2_redundancy_eliminator_bytecodes.c]: If `value` is known to be `None` this can be replaced with `_POP_TOP; LOAD_CONST_BORROW`

**Review by markshannon — CHANGES_REQUESTED** [review]
Looks good in general.

There are a few optimizations that have been missed, and beware the side effects of C API calls.

File: Python/tier2_redundancy_eliminator_bytecodes.c
Comment (Fidget-Spinner) [reply Python/tier2_redundancy_eliminator_bytecodes.c]: I will do the latter, I plan to unify all constants emission later in a separate issue.

File: Python/tier2_redundancy_eliminator_bytecodes.c
Comment (Fidget-Spinner) [reply Python/tier2_redundancy_eliminator_bytecodes.c]: I plan to do this separately in https://github.com/python/cpython/issues/115759

File: Python/tier2_redundancy_eliminator_bytecodes.c
Comment (markshannon) [reply Python/tier2_redundancy_eliminator_bytecodes.c]: "In a *preparatory* issue". That means do it before this, otherwise we will need to go back over all these instructions and redo the optimizations.

If you want to get this PR done first, then use a "super micro op". We will need a cleanup pass either way.

File: Python/optimizer_analysis.c
Comment (Eclips4) [inline Python/optimizer_analysis.c]: Can we adjust these slashes? :)

Comment (gvanrossum) [inline Python/optimizer_analysis.c]: This is an old pep peeve of mine (see also PEP 8 :-):
```suggestion
#define ERROR_IF(COND, LABEL) \
    do { \
        if (COND) { \
            goto LABEL; \
        } \
    } while (0);
```

File: Python/tier2_redundancy_eliminator_bytecodes.c
Comment (gvanrossum) [inline Python/tier2_redundancy_eliminator_bytecodes.c]: Maybe there should be s `get_long_const()` helper which returns `(PyLongObject *)get_const(x)` and also includes the `assert(PyLong_CheckExact())` call, to make code like this more compact. (Same for float.)

File: Python/tier2_redundancy_eliminator_bytecodes.c
Comment (gvanrossum) [inline Python/tier2_redundancy_eliminator_bytecodes.c]: Honestly I'd prefer to break code like this up in two lines:
```suggestion
        res = sym_new_known_type(ctx, &PyBool_Type);
        OUT_OF_SPACE_IF_NULL(res);
```
I think the expression is "don't bury the lede". :-)

File: Python/tier2_redundancy_eliminator_bytecodes.c
Comment (gvanrossum) [inline Python/tier2_redundancy_eliminator_bytecodes.c]: Shouldn't this use `Py_False`?
```suggestion
            REPLACE_OP(this_instr, _POP_TOP_LOAD_CONST_INLINE_BORROW, 0, (uintptr_t)Py_False);
```

File: Python/tier2_redundancy_eliminator_bytecodes.c
Comment (gvanrossum) [inline Python/tier2_redundancy_eliminator_bytecodes.c]: I'd move this up to the top of the block (but after the check for "impossible" cases, if you decide to add that).

File: Python/tier2_redundancy_eliminator_bytecodes.c
Comment (gvanrossum) [inline Python/tier2_redundancy_eliminator_bytecodes.c]: If feels like there should be an `else` block that raises some kind of error if it's a constant but it isn't None. Or maybe there should be a check at the top that it's either unknown or None, and nothing else?

File: Python/tier2_redundancy_eliminator_bytecodes.c
Comment (gvanrossum) [inline Python/tier2_redundancy_eliminator_bytecodes.c]: It just occurred to me that `sym_set_type()` could check that the symbol isn't already known to have a different type that's inconsistent -- that would be either an `assert()` (if we know this can never happen unless there's a bug in the optimizer or trace generator) or it could just abort the optimization process and let the machinery fall back to Tier 1 (if we feel there are legitimate cases that could lead to such an inconsistency).

File: Python/tier2_redundancy_eliminator_bytecodes.c
Comment (Fidget-Spinner) [reply Python/tier2_redundancy_eliminator_bytecodes.c]: Traditionally, the rest of the code sets types after the operation. The thinking is that the types are known after the operation executes.

File: Python/bytecodes.c
Comment (gvanrossum) [inline Python/bytecodes.c]: Maybe there could be an even more specialized opcode that doesn't DECREF its input? That would be handy for the case where you introduced this -- the input is known to be None, so we don't really need to bother with the DECREF of the immortal value.

**Review by gvanrossum — COMMENTED** [review]
One more thing.

File: Python/tier2_redundancy_eliminator_bytecodes.c
Comment (gvanrossum) [reply Python/tier2_redundancy_eliminator_bytecodes.c]: Oh, subtle. I thought that this was an assertion about the input, and it is, but it's not actually known until the op has executed successfully -- it deopts if the input isn't None. So you are right.

File: Python/bytecodes.c
Comment (markshannon) [inline Python/bytecodes.c]: Since there is only one input
```suggestion
            Py_DECREF(pop);
```

File: Python/tier2_redundancy_eliminator_bytecodes.c
Comment (markshannon) [inline Python/tier2_redundancy_eliminator_bytecodes.c]:
```suggestion
        sym_set_const(value, Py_None);
```

**Review by markshannon — COMMENTED** [review]
Two suggestions

File: Python/tier2_redundancy_eliminator_bytecodes.c
Comment (markshannon) [reply Python/tier2_redundancy_eliminator_bytecodes.c]: There is no error, but maybe we should optimize it differently.
Symbolic values aren't real values, so different assumptions apply.

Clearly we need to document this properly, or it is going to keep tripping people up. 
https://github.com/python/cpython/issues/115816

File: Python/tier2_redundancy_eliminator_bytecodes.c
Comment (markshannon) [reply Python/tier2_redundancy_eliminator_bytecodes.c]: See above comment

File: Python/bytecodes.c
Comment (markshannon) [reply Python/bytecodes.c]: The expectation is that the `_POP_TOP` part will be removed in a later pass. 
`LOAD_CONST ; _POP_TOP_LOAD_CONST_INLINE_BORROW` -> `_LOAD_CONST_INLINE_BORROW`

**Review by gvanrossum — APPROVED** [review]
LGTM, let's see if @markshannon agrees.

File: Python/optimizer_analysis.c
Comment (Fidget-Spinner) [inline Python/optimizer_analysis.c]: This is temporary, just ignore this for now. I will remove it in https://github.com/python/cpython/pull/115817.

File: Python/tier2_redundancy_eliminator_bytecodes.c
Comment (markshannon) [inline Python/tier2_redundancy_eliminator_bytecodes.c]: Why is this necessary? Shouldn't `sym_get_const` return `NULL` for anything that isn't a constant?
It does seem rather verbose to have to prefix every `sys_get_const` with a `sys_is_const`.

Comment (Fidget-Spinner) [reply Python/tier2_redundancy_eliminator_bytecodes.c]: In the code:
```C
static inline PyObject *
sym_get_const(_Py_UOpsSymType *sym)
{
    assert(sym_is_const(sym));
    assert(sym->const_val);
    return sym->const_val;
}
```
So we assume there is always something there when there's `sym_get_const`. I think I can just change the contract to return NULL on non constants though.

**Comment by gvanrossum** [comment]
@Fidget-Spinner: It looks like this PR is totally out of date. It adds a new ERROR_IF() macro that should probably be done in a separate cleanup pass. I can do that next (once gh-116062 is merged).

The key part of this PR is to add optimizer specializations for `_TO_BOOL` etc., but it may be simpler to extract those manually and put them in a fresh PR? Or you could do a massive rebase. :-) But I'd still like the ERROR_IF() cleanup to be separated out.

File: Python/tier2_redundancy_eliminator_bytecodes.c
Comment (gvanrossum) [reply Python/tier2_redundancy_eliminator_bytecodes.c]: The new sym_get_const() indeed returns NULL in this case.

File: Python/optimizer_bytecodes.c
Comment (gvanrossum) [inline Python/optimizer_bytecodes.c]: Maybe restore the old (incorrect) code for this and the next two chunks, since my PR gh-116062 will fix this better?

**Review by gvanrossum — APPROVED** [review]
LGTM, with one nit (to make merging of my PR simpler).

File: Python/optimizer_bytecodes.c
Comment (gvanrossum) [inline Python/optimizer_bytecodes.c]: This will need a bottom check (like in gh-#116089). Also the other `sym_set_...()` calls you added below.

**Review by gvanrossum — APPROVED** [review]
Excellent!

**Comment by gvanrossum** [comment]
Since you've addressed Mark's question I don't see a reason to wait for his explicit approval.

**Pull request merged**
